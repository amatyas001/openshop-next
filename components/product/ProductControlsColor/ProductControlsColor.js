import PropTypes from 'prop-types';
import { Text, Select } from '@chakra-ui/core';
import { FaArrowDown } from 'react-icons/fa';
import * as COLORS from '@app/config/colors';
import { ProductShape } from '@app/lib/types';

/**
 * Renders the available colors of a specific item as a select input
 * field or plain text if there is only one option.
 *
 * @see https://amatyas001.github.io/openshop-next/#productcard
 */
export const ProductControlsColor = React.memo(
  (props) => {
    const { product, details, setDetails } = props;
    return Array.isArray(product.color) ? (
      <Select
        placeholder='Select color...'
        icon={FaArrowDown}
        height='50px'
        width='85%'
        fontSize='1.4rem'
        fontFamily='Khand'
        fontWeight='bold'
        mx='auto'
        bg={COLORS.BUTTON.transparent.base}
        color={COLORS.BUTTON.primary.base}
        border='0'
        borderRadius='0'
        value={details.buy.color || ''}
        onChange={(e) => setDetails(e.target.value)}
        _hover={{
          ...COLORS.BUTTON.transparent.hover,
          color: COLORS.BUTTON.primary.hover.bg,
          textDecoration: 'none',
          cursor: 'pointer',
        }}
        _focus={{
          ...COLORS.BUTTON.transparent.hover,
          color: COLORS.BUTTON.primary.hover.bg,
          textDecoration: 'none',
        }}
        _select={{
          ...COLORS.BUTTON.transparent.hover,
          color: COLORS.BUTTON.primary.hover.bg,
          textDecoration: 'none',
        }}
        _active={{
          ...COLORS.BUTTON.transparent.hover,
          color: COLORS.BUTTON.primary.hover.bg,
          textDecoration: 'none',
        }}
      >
        {product.color.map((item) => (
          <option key={item} value={item}>
            {item}
          </option>
        ))}
      </Select>
    ) : (
      <Text
        color={COLORS.BUTTON.primary.base}
        fontFamily='Khand'
        fontSize='1.4rem'
        fontWeight='bold'
        mx='auto'
      >
        {product.color}
      </Text>
    );
  },
  /* istanbul ignore next */
  (prev, next) => prev.details.buy.color === next.details.buy.color
);

ProductControlsColor.propTypes = {
  /**
   * Unique item object
   */
  product: PropTypes.shape({ ...ProductShape }).isRequired,

  /**
   * Unique item object with buying details
   */
  details: PropTypes.shape({
    ...ProductShape,
    buy: PropTypes.shape({
      amount: PropTypes.number.isRequired,
      size: PropTypes.string,
      color: PropTypes.string,
    }),
  }).isRequired,

  /**
   * Set buying details
   */
  setDetails: PropTypes.func.isRequired,
};
