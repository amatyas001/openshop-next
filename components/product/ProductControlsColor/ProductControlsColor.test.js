import { create, act } from 'react-test-renderer';
import { ProductControlsColor } from './ProductControlsColor';

describe('<ProductControlsColor />', () => {
  let tree;

  const mock_product = {
    id: 'mock_id',
    name: 'mock_name',
    price: 10,
    color: ['mock_color_0', 'mock_color_1'],
    amount: 10,
    starrating: 3,
  };

  const mock_details = {
    ...mock_product,
    buy: {
      amount: 0,
    },
  };

  const mock_setDetails = jest.fn();

  const mock_value = 'mock_value';

  beforeAll(() => {
    act(() => {
      tree = create(
        <ProductControlsColor
          product={mock_product}
          details={mock_details}
          setDetails={mock_setDetails}
        />
      );
    });
  });

  it('should render with required props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call handler', () => {
    act(() => {
      tree.root
        .findByType('select')
        .props.onChange({ target: { value: mock_value } });
    });
    expect(mock_setDetails).toHaveBeenCalledTimes(1);
    expect(mock_setDetails).toHaveBeenCalledWith(mock_value);
  });
});
