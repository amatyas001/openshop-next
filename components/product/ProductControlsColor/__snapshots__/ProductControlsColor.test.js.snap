// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductControlsColor /> should render with required props 1`] = `
<div
  className="css-1php3x8"
>
  <select
    className="css-1fbi3l4"
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Select color...
    </option>
    <option
      value="mock_color_0"
    >
      mock_color_0
    </option>
    <option
      value="mock_color_1"
    >
      mock_color_1
    </option>
  </select>
  <div
    className="css-rl6gbg"
  >
    <svg
      aria-hidden="true"
      className="css-k176ff"
      data-custom-icon={true}
      fill="currentColor"
      focusable="false"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
      />
    </svg>
  </div>
</div>
`;
